File: src/Fast2048.py
import numpy as np
from reward_function import ROW_GRADIENT, COL_GRADIENT

class Fast2048:
    move_row_LUT = []
    move_reward_LUT = []
    move_valid_LUT = []

    def __init__(self):
        if not Fast2048.move_row_LUT:
            self.init_LUT()
        self.board = None
        self.max_tile = None
        self.score = None
        self.done = None

        # Curriculum learning parameters
        self.prob_4 = 0.1
        self.p_helpful = 0.0

        self.reset()

    def init_LUT(self):
        for i in range(65536):
            original_row = [(i >> 0) & 0xf, (i >> 4) & 0xf, (i >> 8) & 0xf, (i >> 12) & 0xf]
            row = original_row.copy()

            row = stack_row(row)
            row, reward = merge_row(row)
            row = stack_row(row)

            Fast2048.move_row_LUT.append(row)
            Fast2048.move_reward_LUT.append(reward)

            Fast2048.move_valid_LUT.append(original_row != row)

    def reset(self):
        self.board = np.array([[0 for _ in range(4)]for _ in range(4)])
        self.max_tile = 0
        self.score = 0
        self.done = False
        self.generate_random()
        self.generate_random()
        self.update_values()

    def update_values(self):
        for row in self.board:
            for cell in row:
                self.max_tile = max(self.max_tile, cell)

    def is_move_valid(self, action):
        if action == 3:  # left
            for i in range(4):
                if self.move_valid_LUT[row_to_number(self.board[i])]: return True
        elif action == 1:  # right
            for i in range(4):
                if self.move_valid_LUT[row_to_number(self.board[i][::-1])]: return True
        elif action == 0:  # up
            for i in range(4):
                if self.move_valid_LUT[row_to_number(self.board[:, i])]: return True
        elif action == 2:  # down
            for i in range(4):
                if self.move_valid_LUT[row_to_number(self.board[:, i][::-1])]: return True
        return False

    def generate_random(self):
        # The logic for choosing between a 2 or 4 remains the same
        num = 1 if np.random.random() > self.prob_4 else 2

        empty_cells = np.argwhere(self.board == 0)

        if empty_cells.size == 0:
            return

        # --- REFINED CURRICULUM SPAWN LOGIC ---

        # 1. Rank all empty cells from best to worst
        log_board = np.log2(self.board, out=np.zeros_like(self.board, dtype=np.float32), where=(self.board != 0))
        s1 = np.sum(log_board * ROW_GRADIENT)
        s2 = np.sum(log_board * COL_GRADIENT)
        gradient_to_use = ROW_GRADIENT if s1 >= s2 else COL_GRADIENT

        # Create a list of (gradient_value, position) for each empty cell
        ranked_cells = sorted(
            [(gradient_to_use[r, c], (r, c)) for r, c in empty_cells],
            key=lambda x: x[0]
        )

        # 2. Determine the size of the "good" candidate pool based on p_helpful
        # The pool size shrinks as p_helpful decays from 1.0 to 0.0
        # The formula is len * (1.0 - p_helpful), ensuring at least 1 candidate
        num_candidates = int(np.ceil(len(ranked_cells) * (1.0 - self.p_helpful)))
        num_candidates = max(1, num_candidates)

        # 3. Select the pool of best candidates
        candidate_pool = ranked_cells[:num_candidates]

        # 4. Choose a random position from within that top-tier pool
        chosen_index = np.random.choice(len(candidate_pool))
        chosen_position = candidate_pool[chosen_index][1]

        self.board[chosen_position[0], chosen_position[1]] = num

    def check_done(self):
        res=False
        for i in range(4):
            res|=self.is_move_valid(i)
            if res:
                break
        return not res


    def move(self, direction):
        merge_score=0
        prev=self.board.copy()

        if direction==3: # left
            for i in range(4):
                index= row_to_number(self.board[i])
                merge_score+=self.move_reward_LUT[index]
                self.board[i] = self.move_row_LUT[index]
        elif direction==1: # right
            for i in range(4):
                index= row_to_number(self.board[i][::-1])
                merge_score+=self.move_reward_LUT[index]
                self.board[i] = self.move_row_LUT[index][::-1]
        elif direction==0: # up
            for i in range(4):
                index= row_to_number(self.board[:,i])
                merge_score+=self.move_reward_LUT[index]
                self.board[:,i] = self.move_row_LUT[index]
        elif direction==2: # down
            for i in range(4):
                index= row_to_number(self.board[:,i][::-1])
                merge_score+=self.move_reward_LUT[index]
                self.board[:,i] = self.move_row_LUT[index][::-1]


        self.score+=merge_score


        moved=not np.array_equal(prev, self.board)
        if moved:
            self.generate_random()


        self.update_values()
        self.done=self.check_done()

        return merge_score, self.done, moved

def row_to_number(row):
    return row[0] | row[1]<<4 | row[2]<<8 | row[3]<<12
def stack_row(row):
    for k in range(4):
        for i in range(1, 4):
            if row[i]!=0 and row[i - 1]==0:
                row[i-1]=row[i]
                row[i]=0
    return row
def merge_row(row):
    reward=0
    for i in range(1,4):
        if row[i-1]==row[i] and row[i]!=0:
            row[i-1]+=1
            row[i]=0
            reward+=2**row[i-1]
    return [row,reward]

File: src_cpp/Fast2048.cpp
#pragma once
#include "Fast2048.h"

Fast2048::Fast2048() {
    if (move_row_LUT.empty())
        init_LUT();
    reset();
}

void Fast2048::reset() {
    for (auto &row : board)
        row.fill(0);
    score = 0;
    done = false;
    max_tile = 0;
    generate_random();
    generate_random();
    update_values();
}

std::tuple<int, bool, bool> Fast2048::move(int direction) {
    int merge_score = 0;

    bool moved=is_move_valid(direction);
    if (!moved)
        return {0, done, false};

    if (direction == 3) {
        for (auto &row : board) {
            int index=row_to_number(row);
            merge_score += move_reward_LUT[index];
            row = move_row_LUT[index];
        }
    }
    else if (direction == 1)
    {
        for (auto &row : board) {
            std::reverse(row.begin(), row.end());
            int index=row_to_number(row);
            merge_score += move_reward_LUT[index];
            row = move_row_LUT[index];
            std::reverse(row.begin(), row.end());
        }
    }
    else if (direction == 0)
    {
        for (int col=0;col<4;col++) {
            std::array<int, 4> column;
            for (int row=0;row<4;row++)
                column[row] = board[row][col];
            int index=row_to_number(column);
            merge_score += move_reward_LUT[index];
            column = move_row_LUT[index];
            for (int row=0;row<4;row++)
                board[row][col] = column[row];
        }
    }
    else if (direction == 2)
    {
        for (int col=0;col<4;col++) {
            std::array<int, 4> column;
            for (int row=0;row<4;row++)
                column[row] = board[3-row][col];
            int index=row_to_number(column);
            merge_score += move_reward_LUT[index];
            column = move_row_LUT[index];
            for (int row=0;row<4;row++)
                board[3-row][col] = column[row];
        }
    }
    score += merge_score;


    generate_random();
    update_values();
    done = check_done();

    return {merge_score, done, moved};
}

bool Fast2048::is_move_valid(int direction) const {
    if (direction == 3) {
        for (auto &row : board) {
            if (move_valid_LUT[row_to_number(row)])return true;
        }
    }
    else if (direction == 1)
    {
        for (auto &row : board) {
            std::array<int, 4> reversed_row = row;
            std::reverse(reversed_row.begin(), reversed_row.end());
            if (move_valid_LUT[row_to_number(reversed_row)])return true;
        }
    }
    else if (direction == 0)
    {
        for (int col=0;col<4;col++) {
            std::array<int, 4> column;
            for (int row=0;row<4;row++)
                column[row] = board[row][col];
            if (move_valid_LUT[row_to_number(column)])return true;
        }
    }
    else if (direction == 2)
    {
        for (int col=0;col<4;col++) {
            std::array<int, 4> column;
            for (int row=0;row<4;row++)
                column[row] = board[3-row][col];
            if (move_valid_LUT[row_to_number(column)])return true;
        }
    }
    return false;
}

std::array<std::array<int, 4>, 4> Fast2048::get_board() const {
    return std::array<std::array<int, 4>, 4>(board);
}

void Fast2048::set_board(const std::vector<std::vector<int>> &new_board) {
    board=std::array<std::array<int, 4>, 4>();
    for (int i=0;i<4;i++) {
        for (int j=0;j<4;j++) {
            board[i][j] = new_board[i][j];
        }
    }
    score = 0;
    done = check_done();
    update_values();
}

int Fast2048::get_score() const {
    return score;
}

int Fast2048::get_max_tile() const {
    return max_tile;
}

void Fast2048::init_LUT() {
    for (int i=0;i<65536;i++) {
        std::array<int, 4> original_row,row;
        for (int j=0;j<4;j++) {
            original_row[j] = (i >> (j * 4)) & 0xF;
            row[j] = original_row[j];
        }

        // Stack
        for (int j=0;j<4;j++) {
            for (int k=1;k<4;k++) {
                if (row[k]!=0 && row[k-1] == 0) {
                    std::swap(row[k-1], row[k]);
                }
            }
        }
        // Merge
        int reward = 0;
        for (int j=1;j<4;j++) {
            if (row[j-1]==row[j] && row[j]!=0) {
                row[j-1]++;
                row[j] = 0;
                reward += (1 << row[j-1]);
            }
        }
        // Stack
        for (int j=0;j<4;j++) {
            for (int k=1;k<4;k++) {
                if (row[k]!=0 && row[k-1] == 0) {
                    std::swap(row[k-1], row[k]);
                }
            }
        }

        move_row_LUT.push_back(row);
        move_reward_LUT.push_back(reward);
        move_valid_LUT.push_back(original_row != row);
    }
}

void Fast2048::generate_random() {
    std::vector<std::pair<int,int>> empty_positions;
    for (int i=0;i<4;i++) {
        for (int j=0;j<4;j++) {
            if (board[i][j]==0)
                empty_positions.emplace_back(i,j);
        }
    }

    if (empty_positions.empty())
        return;

    int cell_index = RandomUtil::get().getRandom<int>(0, empty_positions.size() - 1);
    std::pair<int, int> chosen_cell = empty_positions[cell_index];

    double probability = RandomUtil::get().getRandom<double>(0.0, 1.0);

    int new_tile_value = (probability < 0.9) ? 1 : 2;

    board[chosen_cell.first][chosen_cell.second] = new_tile_value;
}

bool Fast2048::check_done() const {
    for (int i=0;i<4;i++) {
        for (int j=0;j<4;j++) {
            if (board[i][j]==0)
                return false;
            if (j<3 && board[i][j]==board[i][j+1])
                return false;
            if (i<3 && board[i][j]==board[i+1][j])
                return false;
        }
    }
    return true;
}

void Fast2048::update_values() {
    for (const auto &row : board) {
        for (const auto &tile : row) {
            if (tile > max_tile)
                max_tile = tile;
        }
    }
}

int Fast2048::row_to_number(const std::array<int, 4> &row) const {
    return row[0] | (row[1] << 4) | (row[2] << 8) | (row[3] << 12);
}

std::vector<std::array<int, 4>> Fast2048::move_row_LUT;
std::vector<int> Fast2048::move_reward_LUT;
std::vector<bool> Fast2048::move_valid_LUT;
